xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/HostSet.java

```diff
@@ -0,0 +1,190 @@
+        return weightedPriorityHealth;
+    }
+
+    int overProvisioningFactor() {
+        return overProvisioningFactor;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("hostsEndpointGroup", hostsEndpointGroup)
+                          .add("healthyHostsEndpointGroup", healthyHostsEndpointGroup)
+                          .add("healthyEndpointGroupPerLocality", healthyEndpointGroupPerLocality)
+                          .add("degradedHostsEndpointGroup", degradedHostsEndpointGroup)
+                          .add("degradedEndpointGroupPerLocality", degradedEndpointGroupPerLocality)
+                          .add("weightedPriorityHealth", weightedPriorityHealth)
+                          .add("overProvisioningFactor", overProvisioningFactor)
+                          .toString();
+    }
+
+    private static WeightedRandomDistributionSelector<LocalityEntry> rebuildLocalityScheduler(
+            Map<Locality, EndpointGroup> eligibleHostsPerLocality,
+            Map<Locality, EndpointGroup> allHostsPerLocality,
+            Map<Locality, Integer> localityWeightsMap,
+            int overProvisioningFactor) {
+        final ImmutableList.Builder<LocalityEntry> localityWeightsBuilder = ImmutableList.builder();
+        for (Locality locality : allHostsPerLocality.keySet()) {
+            final double effectiveWeight =
+                    effectiveLocalityWeight(locality, eligibleHostsPerLocality, allHostsPerLocality,
+                                            localityWeightsMap, overProvisioningFactor);
+            if (effectiveWeight > 0) {
+                localityWeightsBuilder.add(new LocalityEntry(locality, effectiveWeight));
+            }
+        }
+        return new WeightedRandomDistributionSelector<>(localityWeightsBuilder.build());
+    }
+
+    static double effectiveLocalityWeight(Locality locality,
+                                          Map<Locality, EndpointGroup> eligibleHostsPerLocality,
+                                          Map<Locality, EndpointGroup> allHostsPerLocality,
+                                          Map<Locality, Integer> localityWeightsMap,
+                                          int overProvisioningFactor) {
+        final EndpointGroup localityEligibleHosts =
+                eligibleHostsPerLocality.getOrDefault(locality, EndpointGroup.of());
+        final int hostCount = allHostsPerLocality.getOrDefault(locality, EndpointGroup.of()).endpoints().size();
+        if (hostCount == 0) {
+            return 0;
+        }
+        // We compute the availability of a locality via:
+        // (overProvisioningFactor) * (# healthy/degraded of hosts) / (# total hosts)
+        // https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/locality_weight.html
+        final double localityAvailabilityRatio = 1.0 * localityEligibleHosts.endpoints().size() / hostCount;
```

Comment:
nit: Maybe just cast to double?
```suggestion
        final double localityAvailabilityRatio = (double) localityEligibleHosts.endpoints().size() / hostCount;
```